src/renderer/atlas/AtlasEngine.cpp

```diff
@@ -87,89 +87,23 @@ try
-            }
-            else
-            {
-                // scrollOffset/offset = 1
-                // +----------+    +----------+
-                // | xxxxxxxxx|    |          |
-                // |xxxxxxx   | -> | xxxxxxxxx|
-                // |          |    |xxxxxxx   |
-                // +----------+    +----------+
-                const u16 endRow = offset;
-                _api.invalidatedRows.start = 0;
-                _api.invalidatedRows.end = nothingInvalid ? endRow : std::max(_api.invalidatedRows.end, endRow);
-
-                const auto dst = std::copy_n(_p.rows.end() - offset, offset, _p.rowsScratch.begin());
-                std::copy_n(_p.rows.begin(), _p.rows.size() - offset, dst);
-            }
-
-            std::swap(_p.rows, _p.rowsScratch);
-
-            // Scrolling the background bitmap is a lot easier because we can rely on memmove which works
-            // with both forwards and backwards copying. It's a mystery why the STL doesn't have this.
-            {
-                const auto srcOffset = std::max<ptrdiff_t>(0, -offset) * gsl::narrow_cast<ptrdiff_t>(_p.colorBitmapRowStride);
-                const auto dstOffset = std::max<ptrdiff_t>(0, offset) * gsl::narrow_cast<ptrdiff_t>(_p.colorBitmapRowStride);
-                const auto count = _p.colorBitmapDepthStride - std::max(srcOffset, dstOffset);
-                assert(dstOffset >= 0 && dstOffset + count <= _p.colorBitmapDepthStride);
-                assert(srcOffset >= 0 && srcOffset + count <= _p.colorBitmapDepthStride);
-
-                auto src = _p.colorBitmap.data() + srcOffset;
-                auto dst = _p.colorBitmap.data() + dstOffset;
-                const auto bytes = count * sizeof(u32);
-
-                for (size_t i = 0; i < 2; ++i)
-                {
-                    // Avoid bumping the colorBitmapGeneration unless necessary. This approx. further halves
-                    // the (already small) GPU load. This could easily be replaced with some custom SIMD
-                    // to avoid going over the memory twice, but... that's a story for another day.
-                    if (memcmp(dst, src, bytes) != 0)
-                    {
-                        memmove(dst, src, bytes);
-                        _p.colorBitmapGenerations[i].bump();
-                    }
-
-                    src += _p.colorBitmapDepthStride;
-                    dst += _p.colorBitmapDepthStride;
-                }
-            }
+            const u16 begRow = _p.s->cellCount.y + offset;
+            _api.invalidatedRows.start = nothingInvalid ? begRow : std::min(_api.invalidatedRows.start, begRow);
+            _api.invalidatedRows.end = _p.s->cellCount.y;
+        }
+        else
+        {
+            const u16 endRow = offset;
+            _api.invalidatedRows.start = 0;
+            _api.invalidatedRows.end = nothingInvalid ? endRow : std::max(_api.invalidatedRows.end, endRow);
         }
     }

```

Comment:
This moved up from down below (search for `_api.invalidatedRows.end`), because when we do this here, before doing any actual scrolling, we can more easily check for the `MarkAllAsDirty()` case below. That's because this moved code block calculates the range of `invalidatedRows` that would exist after scrolling. The `MarkAllAsDirty()` can then detect if everything's been scrolled off by just checking if the `invalidatedRows` range is the entire viewport.